### PR TITLE
helm: Add Support for Namespace Override

### LINKED ISSUE #20511: CFP: Add a value for overriding the namespace in the helm chart.
## Cilium Feature Proposal
        
Thanks for taking time to make a feature request for Cilium! If you have usage questions, please try the [slack channel](http://slack.cilium.io/) and see the [FAQ](https://github.com/cilium/cilium/issues?utf8=%E2%9C%93&q=is:issue+label:kind/question+) first.

**Is your feature request related to a problem?**

Adding cilium to an umbrella helm chart but installing to a different namespace isn't currently possible.

**Describe the feature you'd like**

Adding `namespaceOverride: ""` to values.yaml (similar to how bitnami makes their helm charts) or how cert-manager adds the 'namespace' value.

bitnami example:
```
## @param namespaceOverride String to fully override common.names.namespace
##
namespaceOverride: ""
```

cert-manager example: 
```
# This namespace allows you to define where the services will be installed into
# if not set then they will use the namespace of the release
# This is helpful when installing cert manager as a chart dependency (sub chart)
namespace: ""
```

**(Optional) Describe your proposed solution**

I think either the way bitnami or cert-manager handles it works well and could be useful.

### PR DESCRIPTION
**Pull request by pr0PM**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/master/USERS.md)
- [x] Thanks for contributing!

This is helpful while deploying charts using gitOps tools like argoCD. Usually a standard among all the charts.

- Patch for the helper.tpl priority class issue
- document generation steps
- spellcheck allowlist
- make install/kubernetes

Fixes: #20511

Signed-off-by: Prateek Mishra [pr0PM@pm.me](mailto:pr0PM@pm.me)

```release-note
Add namespace override support in cilium charts
```

### CODE REVIEW

**Comment by pr0PM** [comment]
Hi @sayboras this is the same [PR](https://github.com/cilium/cilium/pull/23527#issuecomment-1457735408) which broke down last time while fixing inconsistencies in my config.

**Comment by sayboras** [comment]
> this is the same [PR](https://github.com/cilium/cilium/pull/23527#issuecomment-1457735408) which broke down last time while fixing inconsistencies in my config.

Thanks for your help. There is small failure with generated doc checking, can you run the below command and submit the changes in the same commit ?

https://github.com/cilium/cilium/actions/runs/4455626054/jobs/7829681639?pr=24445
```
Make sure to apply your changes to cilium/values.yaml.tmpl and run 'make -C install/kubernetes cilium/values.yaml' to regenerate.
```

**Review by qmonnet — APPROVED** [review]
Looks good to me, thank you!

**Comment by chancez** [comment]
I can't help but feel like this needs a CI test so we can actually verify this works as expected. Do we have a good way for a contributor to add new CI tests?

File: install/kubernetes/cilium/templates/_helpers.tpl [hunk 1]
```diff
@@ -1,3 +1,15 @@
+{{/*
+overridde the release namespace for multi-namespace deployments in combined charts.
```
Comment (squeed) [inline install/kubernetes/cilium/templates/_helpers.tpl]: nit: override

**Comment by pr0PM** [comment]
Personally I wanted to deploy cilium via argo-cd with custom values, using sync waves so that cilium is installed first. 
Can't pass helm values from charts into argo-cd, so we would need to modify the argocd application spec to pass a parameter for namespace.
Would prefer to keep argocd clean as namespace override is something standard across all charts I've come across.

File: install/kubernetes/cilium/templates/_helpers.tpl (hunk 1 above)
Comment (pr0PM) [reply install/kubernetes/cilium/templates/_helpers.tpl]: patched

**Review by chancez — APPROVED** [review]

**Review by squeed — APPROVED** [review]

**Comment by pr0PM** [comment]
Hi just following up, I hope nothing else is pending from my side.

**Comment by squeed** [comment]
@pr0PM looks like this needs a rebase; then it should be good to go.

**Comment by squeed** [comment]
@pr0PM can you rebase and remove the merge commits please?

Separately, I've got a PR to add a tool for testing helm invariants -- https://github.com/cilium/cilium/pull/23881 -- this would be a good thing to add a test for.

**Comment by pr0PM** [comment]
hi can you provide steps on the kind of state we need here, i tried rebase but it didn't help in this case due to normal out of date from main and the switch from master -> main ig.
can't rebase directly on github as the update with rebase button mentions some conflicts.

File: install/kubernetes/cilium/values.yaml.tmpl [hunk 2]
```diff
@@ -4,6 +4,9 @@
 # For example: 1.7, 1.8, 1.9
 # upgradeCompatibility: '1.8'
 
+# -- String to override release namespace for all resources
```
Comment (kaworu) [inline install/kubernetes/cilium/values.yaml.tmpl:7]: Can we add a bit more of context here? e.g. mention that it is useful for tools that are unable to pass other Helm flags than values like argo-cd.

**Review by kaworu — CHANGES_REQUESTED** [review]
Thanks for the patch @pr0PM! This PR needs to be rebased on top of `main`. A couple of files were removed and there is one conflict. Also `git grep Release.Namespace` shows some instances of `.Release.Namespace` that could be updated.

**Comment by nathanjsweet** [comment]
Uh oh. It looks like you created a rebase commit instead of rebasing. You might just want to create a new branch off of main and cherry-pick your commit on top of it.

File: install/kubernetes/cilium/values.yaml.tmpl (hunk 2 above)
Comment (viktor-kurchenko) [reply install/kubernetes/cilium/values.yaml.tmpl:7]: Thanks @pr0PM! Are you going to replace more `.Release.Namespace` in `install/kubernetes/cilium`?

**Comment by joestringer** [comment]
I've marked the PR as draft while you address the feedback on the PR, feel free to click the "Ready for review" button at the bottom of the page to re-request review when you've addressed that. Thanks for your contributions!

**Pull request closed without merging**